homework/pararel_accumulate_async.cpp

```diff
@@ -0,0 +1,116 @@
+#include <iostream>
+#include <thread>
+#include <numeric>
+#include <algorithm>
+#include <future>
+#include <vector>
+#include <chrono>
+#include <functional>
+
+constexpr size_t minimumSize = 200;
+
+struct debugTimers
+{
+	decltype(std::chrono::steady_clock::now()) start3;
+	decltype(start3) stop3;
+	decltype(start3) start4;
+	decltype(start4) stop4;
+};
+
+debugTimers dummy_dt;
+
+template <typename IT, typename T>
+T pararelAccumulate(IT first, IT last, T init, size_t minSizeForThread = minimumSize, debugTimers& dt = dummy_dt ) {
+	const size_t size = std::distance(first, last);
+	if (size < minSizeForThread)
+		return std::accumulate(first, last, init);
+
+	size_t hardwareThread = std::thread::hardware_concurrency();
+
+    if (hardwareThread == 0)
+        hardwareThread = 1;
+
+	const size_t neededThreads = std::min(size / minSizeForThread, hardwareThread);
+	const size_t chunkSize = size / neededThreads;
+	std::vector<T> results(neededThreads);
+    std::vector<std::future<T>> futures(neededThreads);
+
+    auto sumFunction = [](IT first, IT last)
+    {
+        return(std::accumulate(first, last, T{}));
+    };
+
+	dt.start3 = std::chrono::steady_clock::now();
+	auto begin = first;
+	for (size_t i = 0; i < neededThreads; ++i) {
+		auto end = std::next(begin, chunkSize);
+
+        futures[i] = std::async(sumFunction, begin, end);
+		begin = end;
+	}
+	dt.stop3 = std::chrono::steady_clock::now();
+
+	dt.start4 = std::chrono::steady_clock::now();
+    for( int i = 0; i < neededThreads; ++i )
+    {
+        results[i] = futures[i].get();
+    } 
+
+	dt.stop4 = std::chrono::steady_clock::now();
+	return std::accumulate(std::begin(results), std::end(results), init);
```

Comment:
można to też tak przerobić, aby vector results nie był potrzebny. Od razu accumulate na wektorze futures można wywołać. Spróbujesz? :)

Reply:
zrobione